frontend/rust-lib/flowy-core/src/deps_resolve/folder_deps.rs

```diff
@@ -478,3 +483,108 @@ pub fn layout_type_from_view_layout(layout: ViewLayoutPB) -> DatabaseLayoutPB {
     ViewLayoutPB::Document => DatabaseLayoutPB::Grid,
   }
 }
+
+struct FolderIndexStorageImpl(Weak<UserManager>);
+
+impl FolderIndexStorage for FolderIndexStorageImpl {
+  fn search(&self, s: &str, limit: Option<i64>) -> Result<Vec<SearchData>, FlowyError> {
+    let manager = self
```

Comment:
lot of duplicate of this code

Reply:
Yes, I'm not clear on the structure of the crate/dependency flow and how they relate to the state of the application.

When viewed from an object-oriented perspective, a folder contains documents and databases. However, there seems to be no relation between a folder and the documents/databases it contains; there does not appear to be a way for a folder to obtain the contents of its documents/databases.

My assumption is that the different crates were created with boundaries in mind, i.e. folder should not use document/database and vice versa, which limits the implementation of shared functionality in the core package that is the only place that uses all crates.